### PR TITLE
chore: link PRs to linear issues

### LINKED ISSUE #3157: Automatic Linear x Github PR integration
# Problem

When integrating with Linear, PRs now must keep track of two things:
 - GitHub Issues
 - Linear Issues

One would expect Linear to follow this automatically, but it doesn't.

Then, we end up with Github Issues vs. Linear Issues, which is not only noisy but also cumbersome to do manually.

pr-linear

# Desires

 - Have this working automatically out of the box
 - Keep the Linear magic words out of the PR description

# Solution

Implement workflow automation that will:
 - Find all GH issues the PR is closing
 - Find their related Linear issue IDs
 - Modify the PR description, adding Linear magic words as comments (invisible!)

```
<!-- Closes to TS-1, TS-2, TS-N -->
<!-- Part of to TS-5, TS-7, N -->
```

Premisses:
 - GitHub issues must be informed one per line _(predictable parsing)_
 - Linear Comments can contain multiple issues per line _(it doesn't matter)_

Docs:
 - [Github Keywords](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests)
 - [Linear Magic Words](https://linear.app/docs/github?tabs=206cad22125a#link-using-pull-requests) & [Linear API](https://developers.linear.app/docs/graphql/working-with-the-graphql-api/filtering#examples)

# Github Action

The final solution must be a GitHub action that all other repos can reuse; it can go here:
 - https://github.com/FuelLabs/github-actions

### PR DESCRIPTION
**Pull request by nedsalk**
- Closes #3157 

# Summary

- Huge shoutout to @maschad who kindly solved my browser testing problems!

This PR adds a workflow which automatically links PRs to the corresponding linear issues by updating the body of the PR with markdown comments that contain linear-related magic words. The implementation supports all the GitHub magic words (closes, fixes, etc.). If one wants to _relate_ a PR to an issue without setting it as closing for that issue, they can add a "relates to", "related to", or "part of" like this:
- Relates to #ISSUE1
- Related to #ISSUE2
- Part of #ISSUE3

Besides linking issues, I also implemented unlinking issues for cases where somebody erroneously links the wrong issue in GitHub (e.g. wanted to close #ISSUE2 but wrote that it closes #ISSUE1). This isn't automatically handled by linear when the referenced linear issue's identifier is removed from the body of the PR, so I had to call the `linearClient.deleteAttachment` endpoint directly in these cases. 

I initially wanted to implement both linking and unlinking not by modifying the body of the PR but rather by using the API provided by the linear client, but then I realized that doing it like that doesn't add the metadata that gets added when using linear magic words.

# Checklist

- [x] All **changes** are **covered** by **tests** (or not applicable)
- [x] All **changes** are **documented** (or not applicable)
- [x] I **reviewed** the **entire PR** myself (preferably, on GH UI)
- [x] I **described** all **Breaking Changes** (or there's none)

### CODE REVIEW

File: scripts/linear/pr-to-linear-linker.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  );
+
+  const relatedIssuesRegex = /(relates\sto|related\sto|part\sof)\s+#(\d+)/gi;
+  const relatedIssues = [...prBody.matchAll(relatedIssuesRegex)].map(
+    ([, , issueNo]) => `https://github.com/${owner}/${repo}/issues/${issueNo}`
+  );
+  const prUrl = `https://github.com/${owner}/${repo}/pull/${pullNumber}`;
+
+  return { closingIssues, relatedIssues, prUrl };
+}
```
Comment (arboleya) [inline scripts/linear/pr-to-linear-linker.ts:19]: See what you think.

```ts
function parsePr(params: { owner: string; repo: string; prNo: number; prBody: string; }) {
  const { owner, repo, prNo, prBody } = params;

  const prUrl = `https://github.com/${owner}/${repo}/pull/${prNo}`;
  const issueUrl = `https://github.com/${owner}/${repo}/issues/%s`;

  const formatIssueUrl = (r: RegExpExecArray) => issueUrl.replace('%s', r[2]);

  const closingIssuesRegex = /(close(?:s|d)?|fix(?:es|ed)?|resolve(?:s|d))\s#(\d+)/gi;
  const closingIssues = [...prBody.matchAll(closingIssuesRegex)].map(formatIssueUrl);

  const relatedIssuesRegex = /(relate(?:s|d)|part\sof)\s+#(\d+)/gi;
  const relatedIssues = [...prBody.matchAll(relatedIssuesRegex)].map(formatIssueUrl);

  return { closingIssues, relatedIssues, prUrl };
}

```

File: scripts/linear/pr-to-linear-linker.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+export async function prToLinearLinker(params: {
```
Comment (arboleya) [inline scripts/linear/pr-to-linear-linker.ts:21]: Nit.

```suggestion
export async function linkPrToLinear(params: {
```

File: scripts/linear/pr-to-linear-linker.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
+import type { getOctokit } from '@actions/github';
+import type { LinearClient } from '@linear/sdk';
```
Comment (arboleya) [inline scripts/linear/pr-to-linear-linker.ts:2]: Can you please add some comments documenting this file?

File: scripts/tests-find.sh [hunk 4]
```diff
@@ -1,8 +1,11 @@
 #!/bin/bash
 
-ROOT=$(cd "$(dirname "$0")/.."; pwd)
+ROOT=$(
+  cd "$(dirname "$0")/.."
+  pwd
+)
 
-FILES=$(find $ROOT/{apps,packages,internal,templates} -name '*.test.ts')
+FILES=$(find $ROOT/{apps,packages,internal,templates,scripts} -name '*.test.ts')
```
Comment (arboleya) [inline scripts/tests-find.sh:8]: The final solution must be a GitHub action that all other repos can reuse.

Can you please port this PR here?
- https://github.com/FuelLabs/github-actions

Once there, this PR should simply integrate the action from `github-actions` into `fuels-ts`.

Ultimately, this PR will be the reference for integrating the action on other repos.

**Review by arboleya — CHANGES_REQUESTED** [review]

**Comment by nedsalk** [comment]
Superseded by https://github.com/FuelLabs/github-actions/pull/40.

File: scripts/tests-find.sh (hunk 4 above)
Comment (nedsalk) [reply scripts/tests-find.sh:8]: https://github.com/FuelLabs/github-actions/pull/40

**Pull request closed without merging**